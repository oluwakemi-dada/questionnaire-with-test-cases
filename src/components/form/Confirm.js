import React from 'react';
import styled from 'styled-components';
import { useHistory } from 'react-router-dom';
import {
  DetailsWrapper,
  PrimaryHeading,
  BtnWrapper,
  BtnPrev,
  BtnNext,
} from './FormResources';

const Details = styled.div`
  background: rgba(195, 191, 243, 0.173);
  border-radius: 0.5rem;
  padding: 0 2rem;

  > *:last-child {
    border: 0;
  }
`;

const Items = styled.div`
  border-bottom: ${(props) =>
    props.theme === 'light' ? '1px solid #e4dbe7' : '1px solid #545353'};
  padding: 1rem 0;

  div {
    padding: 1rem 0rem;
    display: flex;
    align-items: center;
  }

  div:first-child {
    font-size: 1.4rem;
  }

  div:last-child {
    font-size: 1.55rem;
  }
`;

const Question = styled.div`
  color: ${(props) => (props.theme === 'light' ? '#565454' : '#b0aeae')};
`;

const Answer = styled.div`
  color: ${(props) => (props.theme === 'light' ? '#1c1c1c' : '#e8e5e5')};
`;

const Confirm = ({ prevStep, formData}) => {
  const history = useHistory();
  const currTheme = localStorage.getItem('theme');

  const { name, age, sex } = formData;

  // HANDLERS
  const proceed = (e) => {
    // e.preventDefault();
    history.push('/success');
  };

  const previous = (e) => {
    // e.preventDefault();
    prevStep();
  };

  return (
    <DetailsWrapper>
      <PrimaryHeading className='text-center '>Confirm</PrimaryHeading>
      <Details>
        <Items theme={currTheme}>
          <Question theme={currTheme}>What is your name?</Question>
          <Answer theme={currTheme}>{name ? name : ''}</Answer>
        </Items>
        <Items theme={currTheme}>
          <Question theme={currTheme}>What is your age bracket?</Question>
          <Answer theme={currTheme}>{age ? age : ''}</Answer>
        </Items>
        <Items theme={currTheme}>
          <Question theme={currTheme}>What is your sex?</Question>
          <Answer theme={currTheme}>{sex ? sex : ''}</Answer>
        </Items>
      </Details>
      <BtnWrapper>
        <BtnPrev onClick={(e) => previous(e)}>Back</BtnPrev>
        <BtnNext onClick={(e) => proceed(e)}>Submit</BtnNext>
      </BtnWrapper>
    </DetailsWrapper>
  );
};

export default Confirm;
