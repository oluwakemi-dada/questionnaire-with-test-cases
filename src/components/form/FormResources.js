import styled from 'styled-components';

export const Wrapper = styled.div`
  margin: 0 40rem;

  @media (max-width: 1700px) {
    margin: 0 35rem;
  }

  @media (max-width: 1500px) {
    margin: 0 30rem;
  }

  @media (max-width: 1200px) {
    margin: 0 25rem;
  }

  @media (max-width: 1080px) {
    margin: 0 20rem;
  }

  @media (max-width: 900px) {
    margin: 0 15rem;
  }

  @media (max-width: 768px) {
    margin: 0 10rem;
  }

  @media (max-width: 660px) {
    margin: 0 8rem;
  }

  @media (max-width: 550px) {
    margin: 0 5rem;
  }

  @media (max-width: 450px) {
    margin: 0 2rem;
  }

  @media (max-width: 320px) {
    margin: 0 1rem;
  }
`;

export const DetailsWrapper = styled.div`
  margin: 0 30rem;

  @media (max-width: 1700px) {
    margin: 0 25rem;
  }

  @media (max-width: 1500px) {
    margin: 0 15rem;
  }

  @media (max-width: 1200px) {
    margin: 0 10rem;
  }

  @media (max-width: 1080px) {
    margin: 0 5rem;
  }

  @media (max-width: 900px) {
    margin: 0 5rem;
  }

  @media (max-width: 868px) {
    margin: 0 2rem;
  }
`;

export const PrimaryHeading = styled.h1`
  margin: 2rem 0;
`;

export const FormControl = styled.div`
  display: flex;
  flex-direction: column;
 

  > label {
    margin: 2rem 0 2.5rem 0;
  }

  input,
  select {
    height: 50px;
    border-radius: 10px;
    border: 2px solid #f1f1f1;
    padding: 0 2rem;

    &:focus {
      outline: 0;
      border: 1px solid #5e55ef;
    }
  }

  select {
    background: #fff;
  }
`;

export const Option = styled.div`
  display: flex;
  flex-direction: row;
  align-items: center;

  input {
    margin-right: 2rem;
    width: 2rem;
    cursor: pointer;
    margin-bottom: .2rem;

    &[type='radio'] {
      accent-color: #5e55ef;
    }
  }
`;

export const BtnWrapper = styled.div`
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-between;
  margin-top: 4rem;
  font-size: 1.6rem;
`;

export const BtnPrev = styled.button`
  height: 5rem;
  width: 20rem;
  background: #fff;
  color: #000;
  border: 1px solid #5e55ef;
  border-radius: 10px;
  transition: all 0.3s;
  cursor: pointer;

  @media (max-width: 900px) {
    width: 15rem;
  }

  @media (max-width: 550px) {
    width: 12rem;
  }

  @media (max-width: 375px) {
    width: 11rem;
  }

  &:hover {
    background: #5e55ef;
    color: #fff;
  }
`;

export const BtnNext = styled.button`
  height: 5rem;
  width: 20rem;
  background: #5e55ef;
  border: 1px solid #5e55ef;
  color: #fff;
  border-radius: 10px;
  transition: all 0.3s;
  cursor: pointer;

  @media (max-width: 900px) {
    width: 15rem;
  }

  @media (max-width: 550px) {
    width: 12rem;
  }

  @media (max-width: 375px) {
    width: 11rem;
  }

  &:hover {
    background: #fff;
    color: #000;
  }
`;
