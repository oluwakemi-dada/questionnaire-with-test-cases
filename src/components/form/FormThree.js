import React from 'react';
import Alert from '../layouts/Alert';
import {
  Wrapper,
  PrimaryHeading,
  FormControl,
  BtnWrapper,
  BtnPrev,
  BtnNext,
} from './FormResources';

const FormThree = ({
  nextStep,
  prevStep,
  onChange,
  formData,
  alert,
  setAlertFn,
}) => {
  const { sex } = formData;

  // HANDLERS
  const proceed = (e) => {
    e.preventDefault();
    if (sex === '') {
      setAlertFn('Sex is required');
    } else {
      nextStep();
    }
  };

  const previous = (e) => {
    e.preventDefault();
    prevStep();
  };
  return (
    <Wrapper>
      <PrimaryHeading className='text-center '>Question</PrimaryHeading>

      {alert && <Alert msg={alert} />}

      <FormControl>
        <label htmlFor='sex'>What is your sex?</label>
        <select name='sex' value={sex} onChange={(e) => onChange(e)}>
          <option value=''>Select ...</option>
          <option value='Male'>Male</option>
          <option value='Female'>Female</option>
          <option value='Decline to self-identity'>
            Decline to self-identity
          </option>
        </select>
      </FormControl>

      <BtnWrapper>
        <BtnPrev onClick={(e) => previous(e)}>Back</BtnPrev>
        <BtnNext onClick={(e) => proceed(e)}>Next</BtnNext>
      </BtnWrapper>
    </Wrapper>
  );
};

export default FormThree;
