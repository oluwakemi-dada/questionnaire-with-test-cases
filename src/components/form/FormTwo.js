import React from 'react';
import Alert from '../layouts/Alert';
import {
  Wrapper,
  PrimaryHeading,
  FormControl,
  Option,
  BtnWrapper,
  BtnPrev,
  BtnNext,
} from './FormResources';

const FormTwo = ({
  nextStep,
  prevStep,
  alert,
  setAlertFn,
  selectedOption,
  setSelectedOption,
}) => {
  // HANDLERS
  const proceed = (e) => {
    // e.preventDefault();
    if (selectedOption === '') {
      setAlertFn('Age is required');
    } else {
      nextStep();
    }
  };

  const previous = (e) => {
    // e.preventDefault();
    prevStep();
  };

  const onChangeHandler = (e) => {
    setSelectedOption(e.target.value);
  };

  return (
    <Wrapper>
      <PrimaryHeading className='text-center '>Question</PrimaryHeading>

      {alert && <Alert msg={alert} />}

      <FormControl>
        <label>What is your age bracket?</label>

        <Option>
          <input
            type='radio'
            value='Below 18'
            checked={selectedOption === 'Below 18'}
            onChange={onChangeHandler}
          />
          <label htmlFor='age'>Below 18</label>
        </Option>
        <Option>
          <input
            type='radio'
            value='18 - 22'
            checked={selectedOption === '18 - 22'}
            onChange={onChangeHandler}
          />
          <label htmlFor='age'>18 - 22</label>
        </Option>
        <Option>
          <input
            type='radio'
            value='23 - 30'
            checked={selectedOption === '23 - 30'}
            onChange={onChangeHandler}
          />
          <label htmlFor='age'>23 - 30</label>
        </Option>
        <Option>
          <input
            type='radio'
            value='Above 30'
            checked={selectedOption === 'Above 30'}
            onChange={onChangeHandler}
          />
          <label htmlFor='age'>Above 30</label>
        </Option>
      </FormControl>

      <BtnWrapper>
        <BtnPrev onClick={(e) => previous(e)}>Back</BtnPrev>
        <BtnNext onClick={(e) => proceed(e)}>Next</BtnNext>
      </BtnWrapper>
    </Wrapper>
  );
};

export default FormTwo;
