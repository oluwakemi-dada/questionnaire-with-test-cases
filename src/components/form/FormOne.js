import React from 'react';
import styled from 'styled-components';
import Alert from '../layouts/Alert';
import { Wrapper, PrimaryHeading, FormControl, BtnNext } from './FormResources';

const BtnWrapper = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  margin-top: 4rem;
`;

const FormOne = ({ nextStep, onChange, formData, alert, setAlertFn }) => {
  const { name } = formData;

  // HANDLERS
  const proceed = () => {
    if (name === '') {
      setAlertFn('Name is required');
    } else {
      nextStep();
    }
  };
  return (
    <Wrapper>
      <PrimaryHeading className='text-center '>Question</PrimaryHeading>

      {alert && <Alert msg={alert} />}

      <FormControl>
        <label htmlFor='name'>What is your name?</label>
        <input
          type='text'
          name='name'
          value={name}
          onChange={(e) => onChange(e)}
        />
      </FormControl>

      <BtnWrapper>
        <BtnNext onClick={proceed}>Next</BtnNext>
      </BtnWrapper>
    </Wrapper>
  );
};

export default FormOne;
