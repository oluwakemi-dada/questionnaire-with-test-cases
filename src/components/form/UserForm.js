import React, { useState } from 'react';
import FormOne from './FormOne';
import FormTwo from './FormTwo';
import FormThree from './FormThree';
import Confirm from './Confirm';

const UserForm = () => {
  const [formData, setFormData] = useState({
    name: '',
    age: '',
    sex: '',
  });

  const [selectedOption, setSelectedOption] = useState('');

  const [alert, setAlert] = useState('');

  const [step, setStep] = useState(1);

  // HANDLERS
  const onChange = (e) =>
    setFormData({
      ...formData,
      [e.target.name]: e.target.value,
      age: selectedOption,
    });

  const setAlertFn = (msg) => {
    setAlert(msg);
    setTimeout(() => setAlert(''), 3000);
  };

  // Proceed to next step
  const nextStep = () => {
    setStep(step + 1);
  };

  // Go back to prev step
  const prevStep = () => {
    setStep(step - 1);
  };

  if (step === 1) {
    return (
      <FormOne
        nextStep={nextStep}
        onChange={onChange}
        formData={formData}
        alert={alert}
        setAlertFn={setAlertFn}
      />
    );
  } else if (step === 2) {
    return (
      <FormTwo
        prevStep={prevStep}
        nextStep={nextStep}
        alert={alert}
        setAlertFn={setAlertFn}
        selectedOption={selectedOption}
        setSelectedOption={setSelectedOption}
      />
    );
  } else if (step === 3) {
    return (
      <FormThree
        prevStep={prevStep}
        nextStep={nextStep}
        onChange={onChange}
        formData={formData}
        alert={alert}
        setAlertFn={setAlertFn}
      />
    );
  } else {
    return (
      <Confirm prevStep={prevStep} nextStep={nextStep} formData={formData} />
    );
  }
};

export default UserForm;
