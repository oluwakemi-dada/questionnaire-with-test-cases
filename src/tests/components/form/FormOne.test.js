import React from 'react';
import { shallow } from 'enzyme';
import FormOne from '../../../components/form/FormOne';
import { BtnNext } from '../../../components/form/FormResources';
import { formData } from '../../fixtures/formData';

let nextStep, onChange, setAlertFn, wrapper, alert;
beforeEach(() => {
  alert = '';
  nextStep = jest.fn();
  onChange = jest.fn();
  setAlertFn = jest.fn();
  wrapper = shallow(
    <FormOne
      nextStep={nextStep}
      onChange={onChange}
      formData={formData}
      alert={alert}
      setAlertFn={setAlertFn}
    />
  );
});

test('should render FormOne correctly with default data', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should call setAlertFn when next button is called without filling an input for name', () => {
  const value = '';
  wrapper.find('input').simulate('change', { target: { value } });
  wrapper.find(BtnNext).simulate('click');

  expect(setAlertFn).toHaveBeenLastCalledWith('Name is required');
});

test('should show error message when alert is present', () => {
  wrapper.setProps({
    alert: 'Name is required',
  });

  expect(wrapper).toMatchSnapshot();
});

test('should call nextStep when next button is called with name input', () => {
  wrapper.setProps({
    formData: {
      name: 'Oluwakemi Dada',
      age: '',
      sex: '',
    },
  });
  wrapper.find(BtnNext).simulate('click');
  
  expect(nextStep).toHaveBeenCalled();
});
