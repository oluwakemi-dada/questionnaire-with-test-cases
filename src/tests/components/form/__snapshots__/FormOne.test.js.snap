// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render FormOne correctly with default data 1`] = `
<styled.div>
  <styled.h1
    className="text-center "
  >
    Question
  </styled.h1>
  <styled.div>
    <label
      htmlFor="name"
    >
      What is your name?
    </label>
    <input
      name="name"
      onChange={[Function]}
      type="text"
      value=""
    />
  </styled.div>
  <styled.div>
    <styled.button
      onClick={[Function]}
    >
      Next
    </styled.button>
  </styled.div>
</styled.div>
`;

exports[`should show error message when alert is present 1`] = `
<styled.div>
  <styled.h1
    className="text-center "
  >
    Question
  </styled.h1>
  <Alert
    msg="Name is required"
  />
  <styled.div>
    <label
      htmlFor="name"
    >
      What is your name?
    </label>
    <input
      name="name"
      onChange={[Function]}
      type="text"
      value=""
    />
  </styled.div>
  <styled.div>
    <styled.button
      onClick={[Function]}
    >
      Next
    </styled.button>
  </styled.div>
</styled.div>
`;
